src/libsystemd/sd-event/sd-event.c

```diff
@@ -2206,188 +2208,193 @@ _public_ int sd_event_source_get_enabled(sd_event_source *s, int *m) {
+                break;
 
-                        if (EVENT_SOURCE_WATCH_PIDFD(s)) {
-                                /* yes, we have pidfd */
+        case SOURCE_EXIT:
+                prioq_reshuffle(s->event->exit, s, &s->exit.prioq_index);
+                break;
 
-                                r = source_child_pidfd_register(s, s->enabled);
-                                if (r < 0) {
-                                        s->enabled = SD_EVENT_OFF;
-                                        s->event->n_enabled_child_sources--;
-                                        return r;
-                                }
-                        } else {
-                                /* no pidfd, or something other to watch for than WEXITED */
+        case SOURCE_DEFER:
+        case SOURCE_POST:
+        case SOURCE_INOTIFY:
+                break;
+        default:
+                assert_not_reached("Wut? I shouldn't exist.");
+        }
 
-                                r = event_make_signal_data(s->event, SIGCHLD, NULL);
-                                if (r < 0) {
-                                        s->enabled = SD_EVENT_OFF;
-                                        s->event->n_enabled_child_sources--;
-                                        event_gc_signal_data(s->event, &s->priority, SIGCHLD);
-                                        return r;
-                                }
-                        }
+        source_prioq_reshuffle(s);
 
-                        break;
+        return 0;
+}
 
-                case SOURCE_EXIT:
-                        s->enabled = m;
-                        prioq_reshuffle(s->event->exit, s, &s->exit.prioq_index);
-                        break;
+_public_ int sd_event_source_set_enabled(sd_event_source *s, int m) {
+        int r;
 
-                case SOURCE_DEFER:
-                case SOURCE_POST:
-                case SOURCE_INOTIFY:
-                        s->enabled = m;
-                        break;
+        assert_return(s, -EINVAL);
+        assert_return(IN_SET(m, SD_EVENT_OFF, SD_EVENT_ON, SD_EVENT_ONESHOT), -EINVAL);
+        assert_return(!event_pid_changed(s->event), -ECHILD);
 
-                default:
-                        assert_not_reached("Wut? I shouldn't exist.");
-                }
-        }
+        /* If we are dead anyway, we are fine with turning off
+         * sources, but everything else needs to fail. */
+        if (s->event->state == SD_EVENT_FINISHED)
+                return m == SD_EVENT_OFF ? 0 : -ESTALE;
 
-        if (s->pending)
-                prioq_reshuffle(s->event->pending, s, &s->pending_index);
+        if (s->enabled == m)
+                return 0;
 
-        if (s->prepare)
-                prioq_reshuffle(s->event->prepare, s, &s->prepare_index);
+        s->enabled = m;
+        r = m == SD_EVENT_OFF ? source_disable(s) : source_enable(s, m);
+        if (r < 0)
+                return r;
 
```

Comment:
so if the ratelimit is effective and someone calls sd_event_source_set_enabled(), then this will possibly low-level enable the event source again even though it is still ratelimited...

I am pretty sure we should enable the event source on the low level only if *both* the user-controlled enabled state says so, *and* the ratelimit is not in effect. 

I'd would be great if the source_enable()/source_disable() split could be done in a separate, preparatory commit, it's easier to follow then